### PR TITLE
Use torch.matmul in nn.Linear

### PR DESCRIPTION
**Pull request by colesbury**
This takes advantage of the broadcasting behavior of torch.matmul to
support inputs with more than two dimensions. The extra dimensions are
treated like part of the batch dimension, much like nn.Bottle in Lua
Torch.

There are a few related small performance changes:

 * Addmm computes the gradient in column-major for inputs in
   column-major format
 * Variable.mm calls Addmm in-place with the desired output buffer

### CODE REVIEW

**Comment by colesbury** [comment]
Some micro-benchmark performance here:
https://gist.github.com/colesbury/fcfed34f6ddfde391adee71c67ec24a8

There's not a significant difference in perf with this PR for the sizes I tried. 

One weird thing is that passing in a 3D input can sometimes be faster than the equivalent 2D input (and likely slower for some other cases). The difference is in the `bias.grad` computation. It's because:
* [maybe_unexpand](https://github.com/pytorch/pytorch/blob/8a4eb50ed1ae875bcca6458b232fff289942911e/torch/autograd/_functions/utils.py#L17) is implemented using repeated calls to `sum(0)` for each dimension
* Strangely, this is actually sometimes faster, probably because sum(dim) is not implemented very efficiently. (Too much work in each thread, not enough parallelism)

**Comment by ngimel** [comment]
Yeah, abandoned @csarofeen's PR https://github.com/torch/cutorch/pull/751 tried to address that, I don't remember what happened to it. Most likely reduction kernels are starting a single threadblock.

**Review by soumith — APPROVED** [review]

**Comment by ngimel** [comment]
Microbenchmark 3 (64 x 2408) is much slower for me after this PR (.54 s vs .40 s before), because 3 memcopies are now done, one in each addmm (as opposed to 2 before). Even those 2 should not be needed, and at some point were not needed, I'll have to see how and why they appeared.

**Comment by soumith** [comment]
@ngimel if you can investigate and issue a fix, that'd be awesome. Over the next 1/2 weeks we lost Trevor and Paszke over time off, so things will be slower.

**Comment by ngimel** [comment]
I'm on vacation too:-)

**Comment by soumith** [comment]
haha okay, then i'll look into this. thanks :D

**Comment by ngimel** [comment]
What I've learned yesterday is that all 3 matrix multiplies in linear go through https://github.com/pytorch/pytorch/blob/master/torch/autograd/_functions/blas.py#L19-L26, and even if output and add_matrix data pointers there are the same, after passing through wrappers t and r_ in THC are not the same, and memcopy here https://github.com/pytorch/pytorch/blob/master/torch/lib/THC/generic/THCTensorMathBlas.cu#L245-L249 is triggered. Before PR forward matrix multiply did not trigger the memcopy, both backward did. After PR, all three MMs do. That benchmark looks like it's close to being CPU bound, so nvprof is pretty much useless for timings, can be used only for the order in which kernels are invoked and kernel durations. (I'm getting .8 s instead of .4-.5 when running under nvprof). I've seen the difference only with the latest (8.0.88) cublas, that improved pre-PR time to .4. With 8.0.61 both pre-PR and post-PR is around .5 s (times are on P100).

**Comment by colesbury** [comment]
I'll send a PR to remove the extra copies in the backwards.

The forward didn't previously trigger a memcopy because the bias addition was separate.  

Previously it was:
1) gemm
2) add bias

Now it is:
1) copy bias to output
2) gemm

So there's an extra copy, but one fewer addition. I'm surprised this is slower. We can get old behavior by removing these [three lines](https://github.com/pytorch/pytorch/blob/35ed224d04ba3eb604dd6b8029ae346ecbaad572/torch/nn/functional.py#L536-L538) in nn/functional.py:

**Comment by ngimel** [comment]
I was also surprised this was slower. Extra Memcopy in THC addmm can't explain it, because it is tiny, must be something about the cpu overhead not being hidden with the new order, but hard to say for sure.
